tests/consensus_tests/test_resharding.py

```diff
@@ -2,991 +2,307 @@
-        wait_for_collection_resharding_operations_count(peer_api_uris[0], COLLECTION_NAME, 1)
-        for uri in peer_api_uris:
-            wait_for_collection_resharding_operations_count(uri, COLLECTION_NAME, 0)
 
-        # Assert node shard and point sum count
-        for uri in peer_api_uris:
-            assert check_collection_local_shards_count(uri, COLLECTION_NAME, shard_count)
-            assert check_collection_local_shards_point_count(uri, COLLECTION_NAME, num_points)
+def test_resharding_state_transitions(tmp_path: pathlib.Path):
+    # Bootstrap resharding cluster
+    peer_uris, _ = bootstrap_resharding(tmp_path)
 
-    sleep(1)
+    # Check that we can't (re)apply unexpected resharding state transitions
+    try_requests(peer_uris[0], 400, [
+        start_resharding,
+        lambda peer_uri: start_resharding(peer_uri, direction="down"),
+        commit_write_hashring,
+        finish_resharding,
+    ])
 
-    # Match all points on all nodes exactly
-    data = []
-    for uri in peer_api_uris:
-        r = requests.post(
-            f"{uri}/collections/{COLLECTION_NAME}/points/scroll", json={
-                "limit": 999999999,
-                "with_vectors": True,
-                "with_payload": True,
-            }
-        )
-        assert_http_ok(r)
-        data.append(r.json()["result"])
-    check_data_consistency(data)
-
-
-# Test resharding shard balancing.
-#
-# Sets up a 3 node cluster and a collection with 1 shard and 2 replicas.
-# Performs resharding 7 times and asserts the shards replicas are evenly
-# balanced across all nodes.
-#
-# In this case the replicas are balanced on the second and third node. The first
-# node has all shards because we explicitly set it as shard target all the time.
-@pytest.mark.skip(reason="moving resharding driver to external service")
-def test_resharding_balance(tmp_path: pathlib.Path):
-    assert_project_root()
+    # Commit read hashring
+    resp = commit_read_hashring(peer_uris[0])
+    assert_http_ok(resp)
 
-    num_points = 100
+    # Check that we can't (re)apply unexpected resharding state transitions or abort resharding
+    try_requests(peer_uris[0], 400, [
+        start_resharding,
+        lambda peer_uri: start_resharding(peer_uri, direction="down"),
+        commit_read_hashring,
+        finish_resharding,
+        abort_resharding,
+    ])
+
+    # Commit write hashring
+    resp = commit_write_hashring(peer_uris[0])
+    assert_http_ok(resp)
 
-    # Prevent optimizers messing with point counts
-    env={
-        "QDRANT__STORAGE__OPTIMIZERS__INDEXING_THRESHOLD_KB": "0",
-    }
+    # Check that we can't (re)apply unexpected resharding state transitions or abort resharding
+    try_requests(peer_uris[0], 400, [
+        start_resharding,
+        lambda peer_uri: start_resharding(peer_uri, direction="down"),
+        commit_read_hashring,
+        commit_write_hashring,
+        abort_resharding,
+    ])
+
+    # Finish resharding
```

Comment:
Since we're also checking we cannot reapply, shall we also assert that we cannot finish resharding a second time?